From 6bbe2c97f8323638f29496e29403a7113dcdb96a Mon Sep 17 00:00:00 2001
From: Phillip Susi <psusi@cfl.rr.com>
Date: Fri, 28 Mar 2014 17:09:56 +0000
Subject: Fix assertion failure: head_size <= 63

probe_partition_for_geom() was originally written using PED_ASSERT() to
test for failures in computing the disk geometry based on existing
partition table entries.  That macro has been modified to ignore the
second argument ( which was a return statement ), and instead abort the
program.  This patch rewrites those tests to not use PED_ASSERT() so the
function correctly returns when the computations fail, and the program
can recover gracefully and operate normally.

Last-Update: 2011-03-08
Bug-Ubuntu: http://launchpad.net/bugs/545911

Patch-Name: fix-head-size-assertion.patch
---
 libparted/labels/dos.c | 30 ++++++++++++++++++++----------
 1 file changed, 20 insertions(+), 10 deletions(-)

diff --git a/libparted/labels/dos.c b/libparted/labels/dos.c
index 06466aa..bd6cb79 100644
--- a/libparted/labels/dos.c
+++ b/libparted/labels/dos.c
@@ -646,8 +646,10 @@ probe_partition_for_geom (const PedPartition* part, PedCHSGeometry* bios_geom)
 	if (cyl_size * denum != a_*H - A_*h)
 		return 0;
 
-	PED_ASSERT (cyl_size > 0, return 0);
- 	PED_ASSERT (cyl_size <= 255 * 63, return 0);
+	if (cyl_size <= 0)
+		return 0;
+	if (cyl_size > 255 * 63)
+		return 0;
 
 	if (h > 0)
 		head_size = ( a_ - c * cyl_size ) / h;
@@ -658,18 +660,24 @@ probe_partition_for_geom (const PedPartition* part, PedCHSGeometry* bios_geom)
 		PED_ASSERT (0, return 0);
 	}
 
-	PED_ASSERT (head_size > 0, return 0);
-	PED_ASSERT (head_size <= 63, return 0);
+	if (head_size <= 0)
+		return 0;
+	if (head_size > 63)
+		return 0;
 
 	cylinders = part->disk->dev->length / cyl_size;
 	heads = cyl_size / head_size;
 	sectors = head_size;
 
-	PED_ASSERT (heads > 0, return 0);
-	PED_ASSERT (heads < 256, return 0);
+	if (heads <= 0)
+		return 0;
+	if (heads > 255)
+		return 0;
 
-	PED_ASSERT (sectors > 0, return 0);
-	PED_ASSERT (sectors <= 63, return 0);
+	if (sectors <= 0)
+		return 0;
+	if (sectors > 63)
+		return 0;
 
 	/* Some broken OEM partitioning program(s) seem to have an out-by-one
 	 * error on the end of partitions.  We should offer to fix the
@@ -678,8 +686,10 @@ probe_partition_for_geom (const PedPartition* part, PedCHSGeometry* bios_geom)
 	if (((C + 1) * heads + H) * sectors + S == A)
 		C++;
 
-	PED_ASSERT ((c * heads + h) * sectors + s == a, return 0);
-	PED_ASSERT ((C * heads + H) * sectors + S == A, return 0);
+	if ((c * heads + h) * sectors + s != a)
+		return 0;
+	if ((C * heads + H) * sectors + S != A)
+		return 0;
 
 	bios_geom->cylinders = cylinders;
 	bios_geom->heads = heads;
